cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CertificateRenewer.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.model;
+
+import io.fabric8.kubernetes.api.model.OwnerReference;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.strimzi.certs.CertAndKey;
+import io.strimzi.operator.common.model.Labels;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * CertificateRenewer is a class which takes the cluster CA cert as well as a secret and ensures the secret passed out
+ * is signed by the CA, renewing the certificates if necessary
+ */
+public class CertificateRenewer {
+
+    private final ClusterCa clusterCa;
+    private final String commonName;
+    private final String keyCertName;
+    private final boolean isMaintenanceTimeWindowsSatisfied;
+
+    protected static final Logger log = LogManager.getLogger(CertificateRenewer.class.getName());
+
+    private CertificateRenewer(ClusterCa clusterCa, String commonName, String keyCertName, boolean isMaintenanceTimeWindowsSatisfied) {
+        this.clusterCa = clusterCa;
+        this.commonName = commonName;
+        this.keyCertName = keyCertName;
+        this.isMaintenanceTimeWindowsSatisfied = isMaintenanceTimeWindowsSatisfied;
+    }
+
+    /**
+     * Pre-configure the CertificateRenewer for generating signed certificates
+     *
+     * @param clusterCa
+     * @param commonName the common-name used in te certificate
+     * @param keyCertName the keyCertName used to index data in the secret
+     * @param isMaintenanceTimeWindowsSatisfied
+     */
+    public static CertificateRenewer of(ClusterCa clusterCa, String commonName, String keyCertName, boolean isMaintenanceTimeWindowsSatisfied) {
+        return new CertificateRenewer(clusterCa, commonName, keyCertName, isMaintenanceTimeWindowsSatisfied);
+    }
+
+    /**
+     * Return a new secret signed by the cluster CA
+     *
+     * @param secret the pre-existing secret, may be null
+     * @param secretName the name of the secret to create or modify
+     * @param namespace the namespace to write the secret to
+     * @param labels the labels te secret should contain
+     * @param ownerReference the OwnerReference for the secret
+     */
+    public Secret signedCertificateSecret(Secret secret, String secretName, String namespace, Labels labels, OwnerReference ownerReference) {
+        Map<String, String> signedCertData = new HashMap<>(4);
+        CertAndKey certAndKey = null;
+
+        Optional<CertAndKey> regeneratedCert = maybeRegenerateCertificate(clusterCa, secret, commonName);
+
+        // TODO ensure empty check is fine
```

Comment:
This is slightly differed from the original logic, all tests still pass but in the case where a cert is supposed to be generated but isnt due to an io exception, we now pass through into the else, does this make sense?

Reply:
@ppatierno @tombentley I think you two worked on this IIRC. Any clue about ^^^?

Reply:
Well why is it OK to ignore the IOE when generating a cert? It seems to me that conflating "cert not required" and "cert required but could not be generated due to some error" is likely to result in the cert just quietly never getting regenerated in the error case.